merkle-tree/indexed/src/lib.rs

```diff
@@ -257,32 +330,50 @@ where
         // tree, validating the proof is going to fail.
         let old_low_leaf = low_element.hash::<H>(&low_element_next_value)?;
         let new_low_leaf = new_low_element.hash::<H>(&new_element.value)?;
-        self.merkle_tree.update(
+        let (new_changelog_index, _) = self.merkle_tree.update(
             changelog_index,
             &old_low_leaf,
             &new_low_leaf,
             low_element.index.into(),
             low_leaf_proof,
         )?;
 
-        // Append new element.
-        let new_leaf = new_element.hash::<H>(&low_element_next_value)?;
-        self.merkle_tree.append(&new_leaf)?;
-        let new_low_element_change_log = RawIndexedElement {
+        // Emit changelog entry for low element.
+        let new_low_element = RawIndexedElement {
             value: bigint_to_be_bytes_array::<32>(&new_low_element.value).unwrap(),
             next_index: new_low_element.next_index,
             next_value: bigint_to_be_bytes_array::<32>(&new_element.value)?,
             index: new_low_element.index,
         };
-        self.indexed_changelog.push(new_low_element_change_log);
+        let low_element_changelog_entry = IndexedChangelogEntry {
+            element: new_low_element,
+            proof: low_leaf_proof.clone(),
+            changelog_index: new_changelog_index,
+        };
+        self.indexed_changelog.push(low_element_changelog_entry);
+
+        // Append new element.
+        let mut proof = BoundedVec::with_capacity(self.height);
```

Comment:
Performance improvement could be to just keep proofs of size height - canopy_depth.